### PR TITLE
XS⚠️ ◾ Adding ESLint "no-undefined" rule

### PR DESCRIPTION
**Pull request by muiriswoulfe**
## Summary

This focuses on improving code consistency by replacing the usage of `undefined` with `null` across various files and methods. This change aims to standardize the handling of uninitialized values and improve code readability and maintainability, as well as to align with the ESLint `no-undefined` rule, which highlights the possibility of `undefined` becoming redefined.

Note that unit tests have not eliminated the use of `undefined`, as this is necessary to test some cases.

### Code Consistency Improvements:

* [`eslint.config.mjs`](diffhunk://#diff-9601a8f6c734c2001be34a2361f76946d19a39a709b5e8c624a2a5a0aade05f2R295): Added the rule `"no-undefined": "error"` to enforce the use of `null` instead of `undefined`. [[1]](diffhunk://#diff-9601a8f6c734c2001be34a2361f76946d19a39a709b5e8c624a2a5a0aade05f2R295) [[2]](diffhunk://#diff-9601a8f6c734c2001be34a2361f76946d19a39a709b5e8c624a2a5a0aade05f2R353)

* [`src/task/src/git/octokitGitDiffParser.ts`](diffhunk://#diff-0e959b5630206a4ce8a405c21439f051d4bd44fee10de7c1976ba982365899bbL28-R28): Replaced `undefined` with `null` for the `_firstLineOfFiles` property and updated conditions checking this property. [[1]](diffhunk://#diff-0e959b5630206a4ce8a405c21439f051d4bd44fee10de7c1976ba982365899bbL28-R28) [[2]](diffhunk://#diff-0e959b5630206a4ce8a405c21439f051d4bd44fee10de7c1976ba982365899bbL76-R76) [[3]](diffhunk://#diff-0e959b5630206a4ce8a405c21439f051d4bd44fee10de7c1976ba982365899bbL140-R140)

* [`src/task/src/metrics/inputs.ts`](diffhunk://#diff-e2c9ceff083bc884f43a764417c426aae7403af8406e8f193ce04039ee3be45aL113-R159): Changed method parameters and internal checks from `undefined` to `null` for various input values. [[1]](diffhunk://#diff-e2c9ceff083bc884f43a764417c426aae7403af8406e8f193ce04039ee3be45aL113-R159) [[2]](diffhunk://#diff-e2c9ceff083bc884f43a764417c426aae7403af8406e8f193ce04039ee3be45aL175-R186) [[3]](diffhunk://#diff-e2c9ceff083bc884f43a764417c426aae7403af8406e8f193ce04039ee3be45aL202-R213) [[4]](diffhunk://#diff-e2c9ceff083bc884f43a764417c426aae7403af8406e8f193ce04039ee3be45aL238-R249) [[5]](diffhunk://#diff-e2c9ceff083bc884f43a764417c426aae7403af8406e8f193ce04039ee3be45aL259-R270) [[6]](diffhunk://#diff-e2c9ceff083bc884f43a764417c426aae7403af8406e8f193ce04039ee3be45aL296-R304)

* [`src/task/src/pullRequests/pullRequest.ts`](diffhunk://#diff-0dba5986bb7531816bb1f1c1f0e62596bd4320f449ef8df7c765c8750de5ab0fL83-R87): Updated the `getUpdatedDescription` method to use `null` instead of `undefined` for the `currentDescription` parameter.

* `src/task/src/repos/azureReposInvoker.ts`, `src/task/src/repos/baseReposInvoker.ts`, `src/task/src/repos/gitHubReposInvoker.ts`, `src/task/src/repos/reposInvoker.ts`, `src/task/src/repos/tokenManager.ts`, `src/task/src/runners/azurePipelinesRunnerInvoker.ts`: Replaced various properties and method parameters from `undefined` to `null` and updated related conditions. [[1]](diffhunk://#diff-dd04424eb7f605d4b4584e72cc84c561663aeaad32743966ed30e904a38959e0L44-R44) [[2]](diffhunk://#diff-dd04424eb7f605d4b4584e72cc84c561663aeaad32743966ed30e904a38959e0L157-R157) [[3]](diffhunk://#diff-dd04424eb7f605d4b4584e72cc84c561663aeaad32743966ed30e904a38959e0L340-R340) [[4]](diffhunk://#diff-a221826b34914e756e969871b529aa4d4428c7429ce15f23e537278dd6186c31L35-R42) [[5]](diffhunk://#diff-f940089e58c7285d97fbb04b132d2827a8aaf024094dfcd7ba27351b7ae6b4a3L101-R101) [[6]](diffhunk://#diff-f940089e58c7285d97fbb04b132d2827a8aaf024094dfcd7ba27351b7ae6b4a3L111-R112) [[7]](diffhunk://#diff-f940089e58c7285d97fbb04b132d2827a8aaf024094dfcd7ba27351b7ae6b4a3L152-R153) [[8]](diffhunk://#diff-f940089e58c7285d97fbb04b132d2827a8aaf024094dfcd7ba27351b7ae6b4a3L335-R335) [[9]](diffhunk://#diff-f940089e58c7285d97fbb04b132d2827a8aaf024094dfcd7ba27351b7ae6b4a3L355-R355) [[10]](diffhunk://#diff-f940089e58c7285d97fbb04b132d2827a8aaf024094dfcd7ba27351b7ae6b4a3L375-R382) [[11]](diffhunk://#diff-bc9d5dba7093f3947cd0afd224904d7b841ae0669ce62bcfa31f9262d75ce339L26-R26) [[12]](diffhunk://#diff-bc9d5dba7093f3947cd0afd224904d7b841ae0669ce62bcfa31f9262d75ce339L108-R108) [[13]](diffhunk://#diff-671d489b30768dec900302465ff2ac0602bbbb9db6013cd09ba5841b30c1fb76L59-R61) [[14]](diffhunk://#diff-671d489b30768dec900302465ff2ac0602bbbb9db6013cd09ba5841b30c1fb76L71-R71) [[15]](diffhunk://#diff-671d489b30768dec900302465ff2ac0602bbbb9db6013cd09ba5841b30c1fb76L197-R197) [[16]](diffhunk://#diff-b790de1cbfd28c9e93b23d09d99f5e1c2ab162ee625c0767ebe93ce854dc6c72L53-R62)

These changes collectively enhance the codebase by ensuring a consistent approach to handling uninitialized values, making the code easier to understand and maintain.

### Test Types

- [X] Unit tests
- [X] Manual tests

### Unit Test Coverage

100%

### CODE REVIEW

File: .github/workflows/release-phase-2.yml [hunk 1]
```diff
@@ -18,7 +18,8 @@ jobs:
   release:
     name: Release
     runs-on: ubuntu-latest
-    permissions: {}
+    permissions:
+      contents: write
```
Comment (K-Cully) [inline .github/workflows/release-phase-2.yml:22]: This seems out of scope.

File: eslint.config.mjs [hunk 2]
```diff
@@ -349,6 +350,7 @@ export default tseslint.config(
     rules: {
       "@typescript-eslint/no-confusing-void-expression": "off",
       "@typescript-eslint/no-magic-numbers": "off",
+      "no-undefined": "off",
```
Comment (K-Cully) [inline eslint.config.mjs:353]: why is it "off" here but "error" above?

File: src/task/src/git/octokitGitDiffParser.ts [hunk 3]
```diff
@@ -25,7 +25,7 @@ export default class OctokitGitDiffParser {
   private readonly _axiosWrapper: AxiosWrapper;
   private readonly _logger: Logger;
 
-  private _firstLineOfFiles: Map<string, number> | undefined;
+  private _firstLineOfFiles: Map<string, number> | null = null;
```
Comment (K-Cully) [inline src/task/src/git/octokitGitDiffParser.ts:28]: ??

File: src/task/src/repos/azureReposInvoker.ts [hunk 4]
```diff
@@ -41,7 +41,7 @@ export default class AzureReposInvoker extends BaseReposInvoker {
   private _project = "";
   private _repositoryId = "";
   private _pullRequestId = 0;
-  private _gitApi: IGitApi | undefined;
+  private _gitApi: IGitApi | null = null;
```
Comment (K-Cully) [inline src/task/src/repos/azureReposInvoker.ts:44]: ?

File: src/task/src/repos/gitHubReposInvoker.ts [hunk 5]
```diff
@@ -108,8 +108,8 @@ export default class GitHubReposInvoker extends BaseReposInvoker {
 
     this.initialize();
 
-    let pullRequestComments: GetIssueCommentsResponse | undefined;
-    let fileComments: GetReviewCommentsResponse | undefined;
+    let pullRequestComments: GetIssueCommentsResponse | null = null;
+    let fileComments: GetReviewCommentsResponse | null = null;
```
Comment (K-Cully) [inline src/task/src/repos/gitHubReposInvoker.ts:112]: There are a lot of these "null = null" changes throughout so I'm going to stop commenting on each of them here.
What is the intent with these?

File: src/task/src/runners/azurePipelinesRunnerInvoker.ts [hunk 6]
```diff
@@ -50,18 +50,16 @@ export default class AzurePipelinesRunnerInvoker
...
-  public getEndpointAuthorization(
-    id: string,
-  ): EndpointAuthorization | undefined {
-    const result: taskLib.EndpointAuthorization | undefined =
+  public getEndpointAuthorization(id: string): EndpointAuthorization | null {
+    const result: taskLib.EndpointAuthorization | null =
       this._azurePipelinesRunnerWrapper.getEndpointAuthorization(id, true);
-    if (!result) {
-      return undefined;
+    if (result === null) {
```
Comment (K-Cully) [inline src/task/src/runners/azurePipelinesRunnerInvoker.ts:61]: This looks like a change in behavior.
!result would be true for 0 or "" whereas result === null would be true for those cases

File: eslint.config.mjs (hunk 2 above)
Comment (muiriswoulfe) [reply eslint.config.mjs:353]: I've switched this rule off for the unit tests. We need to be able to use `undefined` within these to exercise all edge cases, so I deemed this was the best solution.

Unfortunately, the diff doesn't make this clear. But this set of rules has a glob to only specify test files.

File: src/task/src/git/octokitGitDiffParser.ts (hunk 3 above)
Comment (muiriswoulfe) [reply src/task/src/git/octokitGitDiffParser.ts:28]: The goal of this is to remove `undefined` where possible as per the rule.

These fields get defined on the first method execution – essentially a lazy initialisation. Previously, they were `undefined` until set. They're now `null` until set, but that requires them to be explicitly set to `null` due to the way TypeScript works (but which probably has no actual impact in terms of generated code).

File: src/task/src/repos/gitHubReposInvoker.ts (hunk 5 above)
Comment (muiriswoulfe) [reply src/task/src/repos/gitHubReposInvoker.ts:112]: See earlier answer

File: src/task/src/repos/azureReposInvoker.ts (hunk 4 above)
Comment (muiriswoulfe) [reply src/task/src/repos/azureReposInvoker.ts:44]: See earlier answer

File: src/task/src/runners/azurePipelinesRunnerInvoker.ts (hunk 6 above)
Comment (muiriswoulfe) [reply src/task/src/runners/azurePipelinesRunnerInvoker.ts:61]: I agree that it is subtly different, so thanks for calling this out. But I believe this should be unproblematic here as the result should never be an integer but a JSON object. After the change, the compiler complained about this, which is the reason I've changed this. I also believe the changes in the wrapper class here means that the `0` case is actually covered anyway, albeit in a different location.

File: .github/workflows/release-phase-2.yml (hunk 1 above)
Comment (muiriswoulfe) [reply .github/workflows/release-phase-2.yml]: That's true. I've moved it to #573 instead now.
